ext/opentelemetry-ext-flask/src/opentelemetry/ext/flask/__init__.py

```diff
@@ -56,113 +57,145 @@ def hello():
-            )
-            token = context.attach(
-                propagators.extract(otel_wsgi.get_header_from_environ, environ)
-            )
+class _InstrumentedFlask(flask.Flask):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
 
-            tracer = trace.get_tracer(__name__, __version__)
-
-            attributes = otel_wsgi.collect_request_attributes(environ)
-            if flask.request.url_rule:
-                # For 404 that result from no route found, etc, we
-                # don't have a url_rule.
-                attributes["http.route"] = flask.request.url_rule.rule
-            span = tracer.start_span(
-                span_name,
-                kind=trace.SpanKind.SERVER,
-                attributes=attributes,
-                start_time=environ.get(_ENVIRON_STARTTIME_KEY),
-            )
-            activation = tracer.use_span(span, end_on_exit=True)
-            activation.__enter__()
-            environ[_ENVIRON_ACTIVATION_KEY] = activation
-            environ[_ENVIRON_SPAN_KEY] = span
-            environ[_ENVIRON_TOKEN] = token
-
-        @self.teardown_request
-        def _teardown_flask_request(exc):
-            activation = flask.request.environ.get(_ENVIRON_ACTIVATION_KEY)
-            if not activation:
-                logger.warning(
-                    "Flask environ's OpenTelemetry activation missing"
-                    "at _teardown_flask_request(%s)",
-                    exc,
-                )
-                return
+        # Single use variable here to avoid recursion issues.
+        self._original_wsgi_ = self.wsgi_app
+        self.wsgi_app = _rewrapped_app(self.wsgi_app)
 
-            if exc is None:
-                activation.__exit__(None, None, None)
-            else:
-                activation.__exit__(
-                    type(exc), exc, getattr(exc, "__traceback__", None)
-                )
-            context.detach(flask.request.environ.get(_ENVIRON_TOKEN))
+        self.before_request(_before_request)
+        self.teardown_request(_teardown_request)
 
 
 class FlaskInstrumentor(BaseInstrumentor):
+    # pylint: disable=protected-access,attribute-defined-outside-init
     """A instrumentor for flask.Flask
 
     See `BaseInstrumentor`
     """
 
-    def __init__(self):
-        super().__init__()
-        self._original_flask = None
-
     def _instrument(self, **kwargs):
         self._original_flask = flask.Flask
         flask.Flask = _InstrumentedFlask
 
+    def instrument_app(self, app):
+        # FIXME remove this protection once some mechanism in the
+        # BaseInstrumentor class is available for these methods.
+        if not self._is_instrumented:
+            app._original_wsgi_app = app.wsgi_app
+            app.wsgi_app = _rewrapped_app(app.wsgi_app)
+
+            app.before_request(_before_request)
+            app.teardown_request(_teardown_request)
+            self._is_instrumented = True
```

Comment:
Does it mean that I cannot instrument two different apps?


```
instrument_app(app1)
instrument_app(app2) <- will fail because is_instrumented was set by the previous call.
```

Reply:
Fixed, moved the check to the app itself.